src/eins-location.c

```diff
@@ -0,0 +1,197 @@
+/* This should be kept in sync with geocode-glib/geocode-glib.h */
+#define LOCATION_ACCURACY_CITY 15000 /* m */
+
+/*
+ * Recorded when startup has finished as defined by the systemd manager DBus
+ * interface. The auxiliary payload contains the parameters sent by the DBus
+ * systemd manager interface as described at
+ * http://www.freedesktop.org/wiki/Software/systemd/dbus/.
+ */
+#define EVENT_USER_LOCATION "abe7af92-6704-4d34-93cf-8f1b46eb09b8"
+
+static void
+on_location_proxy_ready (GObject      *obj,
+                         GAsyncResult *res)
+{
+  GError *error = NULL;
+
+  GeoclueLocation *location = geoclue_location_proxy_new_for_bus_finish (res,
+                                                                         &error);
+  if (error != NULL)
+    {
+      if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        g_critical ("Failed to connect to GeoClue2 service: %s",
+                    error->message);
+      g_error_free (error);
+      return;
+    }
+
+  gdouble latitude = geoclue_location_get_latitude (location);
+  gdouble longitude = geoclue_location_get_longitude (location);
+  gdouble altitude = geoclue_location_get_altitude (location);
+  gdouble accuracy = geoclue_location_get_accuracy (location);
+  /* altitude is set to -1.7976931348623157e+308 if not determined */
+  gboolean altitude_known = (altitude > -1e308);
+
+  GVariant *location_info = g_variant_new ("(ddbdd)", latitude, longitude,
+                                           altitude_known, altitude, accuracy);
+
+  emtr_event_recorder_record_event (emtr_event_recorder_get_default (),
+                                    EVENT_USER_LOCATION, location_info);
+  /* Assumes ownership of location_info */
+
+  g_object_unref (location);
+}
+
+static void
+on_location_updated (GeoclueClient *client,
+                     gchar         *location_path_old,
+                     gchar         *location_path_new)
+{
+  geoclue_location_proxy_new_for_bus (G_BUS_TYPE_SYSTEM,
+                                      G_DBUS_PROXY_FLAGS_NONE,
+                                      "org.freedesktop.GeoClue2",
+                                      location_path_new,
+                                      NULL, /* cancellable */
+                                      (GAsyncReadyCallback) on_location_proxy_ready,
+                                      NULL /* data */);
+}
+
+static void
+on_start_ready (GeoclueClient *client,
+                GAsyncResult  *res)
+{
+  GError *error = NULL;
+
+  if (!geoclue_client_call_start_finish (client, res, &error))
+    {
+      if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        g_critical ("Failed to start GeoClue2 client: %s", error->message);
+      g_error_free (error);
+    }
+}
+
+static void
+on_client_proxy_ready (GObject      *obj,
+                       GAsyncResult *res)
+{
+  GError *error = NULL;
+
+  GeoclueClient *client = geoclue_client_proxy_new_for_bus_finish (res, &error);
```

Comment:
Nothing unrefs the client. I don't know that we need to record location updates as part of this issue, so perhaps you can just unref it after calling `call_start()`, if the `location-updated` signal is not emitted after the callback.
Otherwise it's fine also to just ignore this leak, since it's a singleton object.